ledger/src/blocktree_processor.rs

```diff
@@ -582,27 +658,50 @@ fn process_single_slot(
     bank: &Arc<Bank>,
     last_entry_hash: &Hash,
     opts: &ProcessOptions,
+    recyclers: &VerifyRecyclers,
 ) -> result::Result<(), BlocktreeProcessorError> {
     let slot = bank.slot();
 
     // Fetch all entries for this slot
     let entries = blocktree.get_slot_entries(slot, 0, None).map_err(|err| {
         warn!("Failed to load entries for slot {}: {:?}", slot, err);
-        BlocktreeProcessorError::FailedToLoadEntries
+        BlocktreeProcessorError::from(err)
     })?;
 
-    // If this errors with a fatal error, should mark the slot as dead so
-    // validators don't replay this slot and  see DuplicateSignature errors
-    // later in ReplayStage
-    verify_and_process_slot_entries(&bank, &entries, *last_entry_hash, opts).map_err(|err| {
-        warn!("slot {} failed to verify: {}", slot, err);
-        err
-    })?;
+    // Mark corrupt slots as dead so validators don't replay this slot and
+    // see DuplicateSignature errors later in ReplayStage
+    let result = verify_and_process_slot_entries(&bank, &entries, last_entry_hash, opts, recyclers);
+    if is_result_fatal(&result) {
+        blocktree
+            .set_dead_slot(slot)
+            .expect("Failed to mark slot as dead in blocktree");
+        if let Err(err) = &result {
+            warn!("slot {} failed to verify: {}", slot, err);
+        }
+        result?;
```

Comment:
This change too. Similar concept, we will freeze banks when an error is not fatal